change some empty dicts/lists as default args to None

almarklein opened the pull request:
issues: closes #1983

I searched the code for `={}` and `=[]`. Assuming pep8 is followed and no arguments are specified as `foo = {}`, I should have got them all.

bryevdv commented on bokeh/plotting_helpers.py:
```diff
@@ -469,7 +469,7 @@ def _new_xy_plot(x_range=None, y_range=None, plot_width=None, plot_height=None,
 
 def _handle_1d_data_args(args, datasource=None, create_autoindex=True,
-        suggested_names=[]):
+        suggested_names=None):
```
This may be intentional. I believe it is to create new unique column names when the user does not supply one.

almarklein replied:
How do you mean? I cannot see anything in the source that suggests that?

bryevdv replied:
OK, I think it used to be. Looking closer I think we could actually remove all that `suggested_names` junk entirely, tho that could go in a different PR if you prefer.

almarklein replied:
Sure, I could do that here. The function is never called with the `suggested_names` arguments, so I suppose I can just remove that bit and be done with it?

bryevdv commented on bokeh/properties.py:
```diff
@@ -837,7 +837,8 @@ class List(Seq):
 
-    def __init__(self, item_type, default=[], help=None):
+    def __init__(self, item_type, default=None, help=None):
+        default = default or []
```
I believe these are also intentional, see:
```
    @property
    def default(self):
        if not isinstance(self._default, types.FunctionType):
            return copy(self._default)
        else:
            value = self._default()
            self.validate(value)
            return value
``` 
@mattpap can you comment?

mattpap replied:
This is intentional.

almarklein replied:
I don't understand. The `default` argument represents the default value that this list prop (as attached to a HasProps class) will have, right? There is no *default* default list, or is there? If there is, should it not be documented or implemented differently?

Just for clarity, with the current way of writing it (before this PR):

    from bokeh.properties import HasProps, List, Int
    
    class Foo(HasProps):        
        spam = List(Int)
        eggs = List(Int)
    
    # Modify the default list instance, which is shared by all List properties
    # that did not specify `default` 
    Foo.spam._default.append(3)  
    
    foo1 = Foo()
    print(foo1.spam, foo1.eggs)  # -> [3] [3]

Granted, this is not likely to cause problems anytime soon, but I cannot imagine this is intentional. Although I might be missing a point.

Of all the "fixes" I did in this PR, there is I think one where the situation really led to a bug. The other situations seem currently pretty safe. However, as the code evolves it might at some point lead to very strange bugs. And that's easily avoided by simply never using empty dicts/lists as the default value for an argument.

bryevdv replied:
@mattpap I only have a vague recollection of what is going on here, a slightly more detailed explanation would be appreciated.

bryevdv replied:
ping @mattpap

mattpap replied:
We want to support both `None` and `[]`, because we currently don't have better way to express optional values (I wanted to implement things like `Option(List(Int))`, but haven't find time for this). If you do `default or []`, then:
```
>>> None or []
[]
```
and is what we don't want.

almarklein replied:
Thanks for the explanation. What do you mean with optional values? As I understand it now, you want to be able to write this:

    >> class Foo(HasProps):
    >>     bar = List(default=None)
    >> foo = Foo()
    >> print(foo.bar)
    None

Isn't that strange, since then the value will not be a list?

bryevdv commented on bokeh/query.py:
```diff
@@ -41,7 +41,7 @@ class LEQ(object): pass
 
 
-def match(obj, selector, context={}):
+def match(obj, selector, context=None):
```
`context` may need to be passed explicitly to the recursive calls to `match` now

almarklein replied:
I don't understand well enough what context is to tell (its not documented). Should context be used in the recursion or not? Since context is/was never modified in-place, I don't think this change means that it should be passed to the recursive call now.

almarklein replied:
On the other hand, it seems more natural if context is used during recursion. I'll add it. Let me know if you think it should not.

bryevdv commented on bokeh/sphinxext/bokeh_github.py:
```diff
@@ -34,7 +34,7 @@
 BOKEH_GH = "https://github.com/bokeh/bokeh"
 
-def bokeh_commit(name, rawtext, text, lineno, inliner, options={}, content=[]):
+def bokeh_commit(name, rawtext, text, lineno, inliner, options=None, content=[]):
```
update `options` but not `content`?

almarklein replied:
content is never used.

bryevdv replied:
true, but it seemed to me this PR was about the principal of avoiding potential mistakes. We have to leave the function sig (it's what sphinx expects) and perhaps `content` will be used in the future, so why not change it now?

almarklein replied:
Good point.

bryevdv commented:
@almarklein so I think if the changes in `properties.py` are reverted, this will be good to go?

almarklein commented:
@bryevdv I don't fully understand how "optional values" for props would work, but it looks like we can keep it as-is for now. I reverted the changes for List and Dict and added a note on the reason for its use.

bryevdv commented:
@almarklein the gist is that you can say things like `thing.line_color = None` as a way to specify that there should no line coloring. So, since `None` is a valid value, it's not workable to used it to mean "unset value". In other places we use a filler `_NotSet` class, and we can probably do something similar here in the future.

almarklein commented:
Ah, so should all Prop classes allow None?

mattpap commented:
> Ah, so should all Prop classes allow None?

Currently yes. In future I would like to disallow this and have e.g. `Option(Int)` to allow `None`.

almarklein commented:
Done. This is ready for final review/merge.

bryevdv commented:
Spurious failures on one test, merging.
